pest/src/parser_state.rs

```diff
@@ -42,61 +48,101 @@ pub struct ParserState<'i, R: RuleType> {
-        neg_attempts: vec![],
-        attempt_pos: 0,
-        atomicity: Atomicity::NonAtomic,
-        stack: vec![]
-    };
-
-    if f(&mut state, Position::from_start(input)).is_ok() {
-        let len = state.queue.len();
-        Ok(pairs::new(Rc::new(state.queue), input.as_bytes(), 0, len))
-    } else {
-        state.pos_attempts.sort();
-        state.pos_attempts.dedup();
-        state.neg_attempts.sort();
-        state.neg_attempts.dedup();
-
-        Err(Error::ParsingError {
-            positives: state.pos_attempts,
-            negatives: state.neg_attempts,
-            // All attempted positions were legal.
-            pos: unsafe { position::new(input.as_bytes(), state.attempt_pos) }
-        })
+    let state = ParserState::new(input);
+
+    match f(state) {
+        Ok(s) => {
+            let len = s.queue.len();
+            Ok(pairs::new(Rc::new(s.queue), input.as_bytes(), 0, len))
+        }
+        Err(mut s) => {
+            s.pos_attempts.sort();
+            s.pos_attempts.dedup();
+            s.neg_attempts.sort();
+            s.neg_attempts.dedup();
+
+            Err(Error::ParsingError {
+                positives: s.pos_attempts.clone(),
+                negatives: s.neg_attempts.clone(),
+                // All attempted positions were legal.
+                pos: unsafe { position::new(input.as_bytes(), s.attempt_pos) }
+            })
+        }
     }
 }
 
 impl<'i, R: RuleType> ParserState<'i, R> {
-    /// Wrapper needed to generate tokens.
+
+    /// Allocates a fresh `ParserState` object to the heap and returns the owned `Box`. This `Box`
+    /// will be passed from closure to closure based on the needs of the specified `Parser`.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # use pest;
+    ///
+    /// let input = "";
+    /// let state: Box<pest::ParserState<&str>> = pest::ParserState::new(input);
+    /// ```
+    pub fn new(input: &'i str) -> Box<Self> {
+        Box::new(ParserState {
+            position: Position::from_start(input),
+            queue: vec![],
+            lookahead: Lookahead::None,
+            pos_attempts: vec![],
+            neg_attempts: vec![],
+            attempt_pos: 0,
+            atomicity: Atomicity::NonAtomic,
+            stack: Stack::new()
+        })
+    }
+
+    /// Returns a reference to the current `Position` of the `ParserState`.
+    ///
+    pub fn get_position(&self) -> &Position<'i> {
+        &self.position
+    }
+
+    /// Returns a clone of the current `Position` of the `ParserState`. Ideally, this will likely
+    /// not be needed as manually manipulating the `Position` can lead to a future inconsistent
+    /// state.
+    ///
+    pub fn clone_position(&self) -> Position<'i> {
+        self.position.clone()
+    }
+
+    /// Given a `Position`, replace the current `Position` of the `ParserState` with the new one.
+    ///
+    pub fn with_updated_position(mut self: Box<Self>, position: Position<'i>) -> Box<Self> {
```

Comment:
I would expect a `with` function to take a closure as an argument. Maybe `updated_position` or simply `update_position` would be enough?

Reply:
That's interesting. I wouldn't have necessarily thought a function prefixed with `with` takes a closure. The rationale is that I'm asking the `Box<Self>` to give me a `Box<Self>` with an updated position. `update_position`, at least to me, connotes mutation. Willing to change it though.